### PR TITLE
Draft: Script to generate example SBOM data from a Koji RPM build

### PR DESCRIPTION
**Pull request by twaugh**
Add a script to create example SBOM data from an existing Koji RPM build.

Also run that script to create example data for
* a C++-based RPM with upstream URL, poppler
* a Go-based RPM with vendored deps, openshift-pipelines-client
* a C-based RPM with no upstream URL, openssl

### CODE REVIEW

**Comment by mprpic** [comment]
One more piece that is missing in the purls is the repo ID qualifier that, again, a perfect SBOM would have for each component. I know this is not possible to get from Koji, but for the RPM builds in Konflux it should be available via the rpm lockfile.

**Comment by twaugh** [comment]
> * Looking at the openssl SBOM, the `SPDXRef-Source0` (the `openssl-3.0.7-hobbled.tar.gz` archive) is really the only one we care about I think. That's the source that we want to link to as provenance and identify its location upstream. The rest of the data are Red Hat-specific files that all appear within the SRPM itself (I don't think we need to list them out specifically).

Fixed to only include archives, not plain files.

> * The upstream archive would ideally look something like this:
>   ```json
>     {
>     "SPDXID": "SPDXRef-Source0",
>     "name": "openssl",
>     "versionInfo": "3.0.7",
>     "downloadLocation": "https://openssl.org/source/openssl-3.0.7.tar.gz",
>     "packageFileName": "openssl-3.0.7-hobbled.tar.gz",
>     "externalRefs": [
>       {
>         "referenceCategory": "PACKAGE-MANAGER",
>         "referenceType": "purl",
>         "referenceLocator": "pkg:rpm/redhat/openssl-libs@3.0.7-18.el9_2?arch=aarch64"
>       }
>     ],
>     "checksums": [
>       {
>         "algorithm": "SHA256",
>         "checksumValue": "4105046836812ed422922f851a57500118a99cc0f009b7eff2b3436110393377"
>       }
>     ]
>   },
>   ```

I think maybe we don't want the externalRefs there? I updated it to match all of this with that exception.

>   I know the script can't divine the upstream URL but maybe we could hardcode it so it's included in the example, since the example should show a "perfect" SBOM.

Yes, done.

>   We would also want to specify the rest of the metadata for each package (namely `copyrightText`, `licenseConcluded`, `licenseDeclared`, `"supplier": "Organization: Red Hat"`), but if the purpose of this example is to show the composition, we can ignore them for now.

I added supplier. For license information it starts to get more difficult, as we ideally need SPDX license identifiers there, which we don't use consistently in RPM spec files. I'll see if I can do something here.

> * For the binary RPMs, `versionInfo` should not include the arch; we specify it as a qualifier in the purl too, not part of the version.

Fixed.

> * Can you remind me why we opted to use the `SPDXRef-DOCUMENT DESCRIBES` relationship for the SRPM component (expected) and also all the binary RPMs (unsure why)? I think the purpose of that relationship is to simply identify some sort of a root element (package in our case) that the document refers to, which I think in our case should just be the SRPM.

I see. Removed those.

> For the structure of the repo: can you add all these examples under an `sbom/` directory? I want to have other directories for purl and vex and then build some sort of a docs site out of the .md files in each directory. mkdocs looks pretty simple to use for this.

Do you mean `sbom/examples/rpm`? Moved.

**Comment by twaugh** [comment]
> One more piece that is missing in the purls is the repo ID qualifier that, again, a perfect SBOM would have for each component. I know this is not possible to get from Koji, but for the RPM builds in Konflux it should be available via the rpm lockfile.

I don't think an RPM lockfile will help here. These examples are not describing build dependencies (at least, not yet) but but RPMs themselves. The repository_id we'd need to include in the purl is that relating to the _release_ location. If we have that at all, it would need to come from the ReleasePlanAdmission.

**Comment by mprpic** [comment]
> I don't think an RPM lockfile will help here. These examples are not describing build dependencies (at least, not yet) but but RPMs themselves. The repository_id we'd need to include in the purl is that relating to the _release_ location. If we have that at all, it would need to come from the ReleasePlanAdmission.

Mmm, you're right. I was getting it confused with RPMs installed inside of a container, where we know where those came from.

For RPM SBOMs themselves, I guess we're hitting the limits of what can generated at build time and what additional data can be added at release time. This is the same discussion as adding product-specific metadata into SBOMs that we discussed earlier today on a meeting.

I wonder if we need some standardized approach on adding release-specific data to a build-time SBOM to make it a release-time SBOM ([SBOM type docs](https://www.cisa.gov/sites/default/files/2023-04/sbom-types-document-508c.pdf) I think call it _Deployed_).

**Review by mprpic — COMMENTED** [review]
In the openshift-pipelines-client example, can the syft output can be post-processed to remove the cpe23 refs and just leave purl. We don't really use CPE 2.3 anywhere, especially for component identification, and it feels duplicative to the purls that we do care about.

That same example also contains this package:
```
{
  "name": "actions/checkout",
  "SPDXID": "SPDXRef-Package-github-action-actions-checkout-ad87d765e374acd4",
  "versionInfo": "v2",
  "supplier": "NOASSERTION",
  "downloadLocation": "NOASSERTION",
  "filesAnalyzed": false,
  "sourceInfo": "acquired package info from GitHub Actions workflow file or composite action file: /.github/workflows/codeql-analysis.yml",
  "licenseConcluded": "NOASSERTION",
  "licenseDeclared": "NOASSERTION",
  "copyrightText": "NOASSERTION",
  "externalRefs": [
    {
      "referenceCategory": "SECURITY",
      "referenceType": "cpe23Type",
      "referenceLocator": "cpe:2.3:a:actions\\/checkout:actions\\/checkout:v2:*:*:*:*:*:*:*"
    },
    {
      "referenceCategory": "PACKAGE-MANAGER",
      "referenceType": "purl",
      "referenceLocator": "pkg:github/actions/checkout@v2"
    }
  ]
},
```
but I think that's more of a build dependency rather than actually shipped as part of the component, right? Is there a way to exclude, or call it out at some docs at least?

File: sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json
Comment (twaugh) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:60]: The content seems to originate from here: https://gitlab.freedesktop.org/poppler/test
The spec file says:
```
# git archive --prefix test/
Source1: %{name}-test-%{test_date}.tar.xz
```

File: sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json
Comment (twaugh) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:45]: Added these where a download URL is available.

**Comment by twaugh** [comment]
> In the openshift-pipelines-client example, can the syft output can be post-processed to remove the cpe23 refs and just leave purl. We don't really use CPE 2.3 anywhere, especially for component identification, and it feels duplicative to the purls that we do care about.

Filtered, so that we only include purl references.

> That same example also contains this package:
> 
> ```
> {
[...]
>     {
>       "referenceCategory": "PACKAGE-MANAGER",
>       "referenceType": "purl",
>       "referenceLocator": "pkg:github/actions/checkout@v2"
>     }
>   ]
> },
> ```
> 
> but I think that's more of a build dependency rather than actually shipped as part of the component, right? Is there a way to exclude, or call it out at some docs at least?

Turned off the GitHub Actions catalogers in the syft invocation.

File: sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json
Comment (twaugh) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:45]: Actually, a question about this. For openssl we now include this:
```
    {
      "SPDXID": "SPDXRef-Source0",
      "name": "openssl",
      "versionInfo": "3.0.7",
      "downloadLocation": "https://openssl.org/source/openssl-3.0.7.tar.gz",
      "packageFileName": "openssl-3.0.7-hobbled.tar.gz",
      "checksums": [
        {
          "algorithm": "SHA256",
          "checksumValue": "4105046836812ed422922f851a57500118a99cc0f009b7eff2b3436110393377"
        }
      ],
      "externalRefs": [
        {
          "referenceCategory": "PACKAGE-MANAGER",
          "referenceType": "purl",
          "referenceLocator": "pkg:generic/openssl@3.0.7?download_url=https://openssl.org/source/openssl-3.0.7.tar.gz&checksum=sha256:4105046836812ed422922f851a57500118a99cc0f009b7eff2b3436110393377"
        }
      ]
    },
```

But of course the checksum is wrong in the purl. We don't have the original archive to checksum.

Technically it should probably be:

```
  "packages": [
    ...
    {
      "SPDXID": "SPDXRef-openssl-3.0.7",
      "name": "openssl",
      "versionInfo": "3.0.7",
      "downloadLocation": "https://openssl.org/source/openssl-3.0.7.tar.gz",
      "externalRefs": [
        {
          "referenceCategory": "PACKAGE-MANAGER",
          "referenceType": "purl",
          "referenceLocator": "pkg:generic/openssl@3.0.7?download_url=https://openssl.org/source/openssl-3.0.7.tar.gz"
        }
      ]
    },
    {
      "SPDXID": "SPDXRef-Source0",
      "name": "openssl",
      "versionInfo": "3.0.7",
      "packageFileName": "openssl-3.0.7-hobbled.tar.gz",
      "checksums": [
        {
          "algorithm": "SHA256",
          "checksumValue": "4105046836812ed422922f851a57500118a99cc0f009b7eff2b3436110393377"
        }
      ]
    },
  ...
  ],
  "relationships": [
    ...
    {
      "spdxElementId": "SPDXRef-Source0",
      "relationshipType": "GENERATED_FROM",
      "relatedSpdxElement": "SPDXRef-openssl-3.0.7"
    },
    ...
  ]
```

WDYT?

Comment (mprpic) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:45]: Mmm, good catch. I wonder if we care about adding a separate node for `openssl-3.0.7-hobbled.tar.gz` then. It's the internally cached version of what goes into the SRPM that nobody can access really, but the SRPM contains that same content (essentially). Maybe the relationship should go straight from the openssl.org archive to the SRPM (via "contains")?

Comment (twaugh) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:45]: Discussing this with tmlcoch I now think this would be handled with a "mid-stream" style upstream repository. I've re-factored the examples to mock that.

**Comment by twaugh** [comment]
Anything needed here before merge?

File: sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json
Comment (mprpic) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:45]: But then we're losing the relationship to the original upstream component, no? Or is the expectation that the midstream repo would itself contain an SBOM that would document the relationship to its upstream counterpart?

Comment (twaugh) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:45]: That's true. (Even if the midstream repo had such an SBOM, we'd also need a process to ingest _that_.)

For the use-case of mapping an upstream project version to all built artifacts containing that version, what do you think will embody "upstream project version"? Will it be package name and versionInfo, or will it be purl?

Comment (mprpic) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:46]: > For the use-case of mapping an upstream project version to all built artifacts containing that version, what do you think will embody "upstream project version"? Will it be package name and versionInfo, or will it be purl?

Ideally both, but it's really the location of where it came from that is important. If tomorrow I decide to publish an `openssl-3.0.7.tar.gz` archive on https://foo.xyz, it will be the purl's `download_url` that sets differentiates it from the actual, official OpenSSL tar ball.

I can also imagine that once we have more of this data, you could build an allowlist of upstreams so that if an RPM includes sources from e.g. openssl.org in one build, it will be flagged as suspicious if the next time you try to build it it tries to pre-fetch sources from foo.xyz.

Comment (twaugh) [reply sbom/examples/rpm/poppler-21.01.0-19.el9-sbom.json:46]: I've pushed a new change which models openssl with a midstream repository that reports the origin URL and digest in a way that we can pull it into our SBOM.

File: sbom/examples/rpm/openssl-3.0.7-18.el9_2-sbom.json
Comment (mprpic) [inline sbom/examples/rpm/openssl-3.0.7-18.el9_2-sbom.json:47]: I think this is just missing a purl, otherwise LGTM!

```
pkg:generic/openssl@3.0.7?download_url=[https://github.com/(RH openssl midstream repo)/archive/refs/tags/3.0.7.tar.gz](https://openssl.org/source/old/3.0/openssl-3.0.7.tar.gz)&checksum=sha256:83049d042a260e696f62406ac5c08bf706fd84383f945cf21bd61e9ed95c396e
```

**Review by mprpic — APPROVED** [review]

File: sbom/examples/rpm/openssl-3.0.7-18.el9_2-sbom.json
Comment (twaugh) [reply sbom/examples/rpm/openssl-3.0.7-18.el9_2-sbom.json:54]: Good catch, added.

Comment (mprpic) [reply sbom/examples/rpm/openssl-3.0.7-18.el9_2-sbom.json:54]: Feel free to merge this if you're happy with it!

**Pull request merged**
